File: ws_zxz_chapt2/src/py02_service/py02_service/demo02_client.py
import rclpy
from rclpy.node import Node
from rclpy.logging import get_logger
from base_interfaces.srv import AddTwoInts
import sys

class MinimalClient(Node):

    def __init__(self):
        super().__init__('minimal_client_py')
        self.get_logger().info("客户端已创建")

        self.client = self.create_client(AddTwoInts, 'add_ints')
        
        while not self.client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('服务连接中，请稍候...')
            
        self.req = AddTwoInts.Request()

    def send_request(self):
        # sys获取参数
        self.req.num1 = int(sys.argv[1])
        self.req.num2 = int(sys.argv[2])
        
        self.future = self.client.call_async(self.req)


def main():
    if len(sys.argv) != 3:
        get_logger("rclpy").error("请输入两个整型参数")
        return
          
    rclpy.init()

    client_node = MinimalClient()

    client_node.send_request()

    rclpy.spin_until_future_complete(client_node, client_node.future)

    try:
        response = client_node.future.result()
        client_node.get_logger().info(f"响应结果:{client_node.req.num1}+{client_node.req.num2}={response.sum}") # type: ignore
    except Exception as e:
        client_node.get_logger().info(
            '服务请求失败： %r' % (e,))


    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ws_zxz_chapt2/src/cpp02_service/src/demo02_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "base_interfaces/srv/add_two_ints.hpp"

using base_interfaces::srv::AddTwoInts;
using namespace std::chrono_literals;

class MyClient: public rclcpp::Node 
{
public:
  MyClient(std::string name)
  : Node(name)
  , cnt(0)
  {
    client_ = this->create_client<AddTwoInts>("add_server");
    RCLCPP_INFO(this->get_logger(),"client create");
  }

  bool connectToServer()
  {
    while(!client_->wait_for_service(1s)){
      if(!rclcpp::ok()){
        RCLCPP_INFO(rclcpp::get_logger("debug"),"program abort");
        return false;
      }
      RCLCPP_INFO(rclcpp::get_logger("debug"),"connecting...");
    }
    RCLCPP_INFO(rclcpp::get_logger("debug"),"connected");
    return true;
  }

  rclcpp::Client<AddTwoInts>::FutureAndRequestId send_request(int num1, int num2)
  {
    auto request = std::make_shared<AddTwoInts::Request>();
    request->num1 = num1;
    request->num2 = num2;
    return client_->async_send_request(request);
  }

private:
  int cnt;
  rclcpp::Client<AddTwoInts>::SharedPtr client_;
};

int main(int argc, char* argv[])
{
  if(argc != 3){
    RCLCPP_ERROR(rclcpp::get_logger("debug"),"input two argument");
    return -1;
  }

  rclcpp::init(argc, argv);
  
  auto node = std::make_shared<MyClient>("client_node");

  if(!node->connectToServer()){
    RCLCPP_INFO(rclcpp::get_logger("debug"),"connect server fail");
    return -2;
  }

  auto future = node->send_request(atoi(argv[1]),atoi(argv[2]));

  if(rclcpp::spin_until_future_complete(node,future) == rclcpp::FutureReturnCode::SUCCESS){
    RCLCPP_INFO(rclcpp::get_logger("debug"),"response success,sum=%d",future.get()->sum);
  }
  else{
    RCLCPP_INFO(rclcpp::get_logger("debug"),"response fail");
  }

  rclcpp::shutdown();
  
  return 0;
}